// purs-loader/lib/index.js
'use strict';

var debug_ = require('debug');

var debug = debug_('purs-loader');

var debugVerbose = debug_('purs-loader:verbose');

var loaderUtils = require('loader-utils');

var Promise = require('bluebird');

var path = require('path');

var PsModuleMap = require('./purs-module-map');

var compile = require('./compile');

var bundle = require('./bundle');

var ide = require('./ide');

var toJavaScript = require('./to-javascript');

var dargs = require('./dargs');

var spawn = require('cross-spawn').sync;

var eol = require('os').EOL;

module.exports = function purescriptLoader(source, map) {
  this.cacheable && this.cacheable();

  var webpackConfig = this.options;

  var cache = webpackConfig.purescriptLoaderCache = webpackConfig.purescriptLoaderCache || {
    rebuild: false,
    deferred: [],
    bundleModules: [],
    ideServer: null,
    psModuleMap: null,
    warnings: [],
    errors: [],
    compilationStarted: false,
    compilationFinished: false,
    installed: false,
    srcOption: []
  };

  var callback = this.async();

  var loaderOptions = loaderUtils.getOptions(this) || {};

  var srcOption = function (pscPackage) {
    var srcPath = path.join('src', '**', '*.purs');

    var bowerPath = path.join('bower_components', 'purescript-*', 'src', '**', '*.purs');

    if (cache.srcOption.length > 0) {
      return cache.srcOption;
    } else if (pscPackage) {
      var pscPackageCommand = 'psc-package';

      var pscPackageArgs = ['sources'];

      var loaderSrc = loaderOptions.src || [srcPath];

      debug('psc-package %s %o', pscPackageCommand, pscPackageArgs);

      var cmd = spawn(pscPackageCommand, pscPackageArgs);

      if (cmd.error) {
        throw new Error(cmd.error);
      } else if (cmd.status !== 0) {
        var error = cmd.stdout.toString();

        throw new Error(error);
      } else {
        var result = cmd.stdout.toString().split(eol).filter(function (v) {
          return v != '';
        }).concat(loaderSrc);

        debug('psc-package result: %o', result);

        cache.srcOption = result;

        return result;
      }
    } else {
      var _result = loaderOptions.src || [bowerPath, srcPath];

      cache.srcOption = _result;

      return _result;
    }
  }(loaderOptions.pscPackage);

  var options = Object.assign({
    context: webpackConfig.context,
    psc: null,
    pscArgs: {},
    pscBundle: null,
    pscBundleArgs: {},
    pscIdeClient: null,
    pscIdeClientArgs: {},
    pscIdeServer: null,
    pscIdeServerArgs: {},
    pscIde: false,
    pscIdeColors: loaderOptions.psc === 'psa',
    pscPackage: false,
    bundleOutput: 'output/bundle.js',
    bundleNamespace: 'PS',
    bundle: false,
    warnings: true,
    watch: false,
    output: 'output',
    src: []
  }, loaderOptions, {
    src: srcOption
  });

  if (!cache.installed) {
    debugVerbose('installing purs-loader with options: %O', options);

    cache.installed = true;

    // invalidate loader cache when bundle is marked as invalid (in watch mode)
    this._compiler.plugin('invalid', function () {
      debugVerbose('invalidating loader cache');

      cache = webpackConfig.purescriptLoaderCache = {
        rebuild: options.pscIde,
        deferred: [],
        bundleModules: [],
        ideServer: cache.ideServer,
        psModuleMap: cache.psModuleMap,
        warnings: [],
        errors: [],
        compilationStarted: false,
        compilationFinished: false,
        installed: cache.installed,
        srcOption: []
      };
    });

    // add psc warnings to webpack compilation warnings
    this._compiler.plugin('after-compile', function (compilation, callback) {
      cache.warnings.forEach(function (warning) {
        compilation.warnings.push(warning);
      });

      cache.errors.forEach(function (error) {
        compilation.errors.push(error);
      });

      callback();
    });
  }

  var psModuleName = PsModuleMap.matchModule(source);

  var psModule = {
    name: psModuleName,
    load: function load(js) {
      return callback(null, js);
    },
    reject: function reject(error) {
      return callback(error);
    },
    srcPath: this.resourcePath,
    srcDir: path.dirname(this.resourcePath),
    jsPath: path.resolve(path.join(options.output, psModuleName, 'index.js')),
    options: options,
    cache: cache,
    emitWarning: function emitWarning(warning) {
      if (options.warnings && warning.length) {
        cache.warnings.push(warning);
      }
    },
    emitError: function emitError(error) {
      if (error.length) {
        cache.errors.push(error);
      }
    }
  };

  debug('loading %s', psModule.name);

  if (options.bundle) {
    cache.bundleModules.push(psModule.name);
  }

  if (cache.rebuild) {
    var connect = function connect() {
      if (!cache.ideServer) {
        cache.ideServer = true;

        return ide.connect(psModule).then(function (ideServer) {
          cache.ideServer = ideServer;
          return psModule;
        }).then(ide.loadWithRetry).catch(function (error) {
          if (cache.ideServer.kill) {
            debug('ide failed to initially load modules, stopping the ide server process');

            cache.ideServer.kill();
          }

          cache.ideServer = null;

          return Promise.reject(error);
        });
      } else {
        return Promise.resolve(psModule);
      }
    };

    var rebuild = function rebuild() {
      return ide.rebuild(psModule).then(function () {
        return toJavaScript(psModule).then(psModule.load).catch(psModule.reject);
      }).catch(function (error) {
        if (error instanceof ide.UnknownModuleError) {
          // Store the modules that trigger a recompile due to an
          // unknown module error. We need to wait until compilation is
          // done before loading these files.

          cache.deferred.push(psModule);

          if (!cache.compilationStarted) {
            cache.compilationStarted = true;

            return compile(psModule).then(function () {
              cache.compilationFinished = true;
            }).then(function () {
              return PsModuleMap.makeMap(options.src).then(function (map) {
                debug('rebuilt module map after unknown module forced a recompilation');

                cache.psModuleMap = map;
              });
            }).then(function () {
              return Promise.map(cache.deferred, function (psModule) {
                return ide.load(psModule).then(function () {
                  return toJavaScript(psModule);
                }).then(psModule.load);
              });
            }).catch(function (error) {
              cache.deferred[0].reject(error);

              cache.deferred.slice(1).forEach(function (psModule) {
                psModule.reject(new Error('purs-loader failed'));
              });
            });
          } else {
            // The compilation has started. We must wait until it is
            // done in order to ensure the module map contains all of
            // the unknown modules.
          }
        } else {
          debug('ide rebuild failed due to an unhandled error: %o', error);

          psModule.reject(error);
        }
      });
    };

    connect().then(rebuild);
  } else if (cache.compilationFinished) {
    debugVerbose('compilation is already finished, loading module %s', psModule.name);

    toJavaScript(psModule).then(psModule.load).catch(psModule.reject);
  } else {
    // The compilation has not finished yet. We need to wait for
    // compilation to finish before the loaders run so that references
    // to compiled output are valid. Push the modules into the cache to
    // be loaded once the complation is complete.

    cache.deferred.push(psModule);

    if (!cache.compilationStarted) {
      cache.compilationStarted = true;

      compile(psModule).then(function () {
        cache.compilationFinished = true;
      }).then(function () {
        if (options.bundle) {
          return bundle(options, cache.bundleModules);
        }
      }).then(function () {
        return PsModuleMap.makeMap(options.src).then(function (map) {
          debug('rebuilt module map after compilation');

          cache.psModuleMap = map;
        });
      }).then(function () {
        return Promise.map(cache.deferred, function (psModule) {
          return toJavaScript(psModule).then(psModule.load);
        });
      }).catch(function (error) {
        cache.deferred[0].reject(error);

        cache.deferred.slice(1).forEach(function (psModule) {
          psModule.reject(new Error('purs-loader failed'));
        });
      });
    } else {
      // The complation has started. Nothing to do but wait until it is
      // done before loading all of the modules.
    }
  }
};
